src/plugins/markdown.ts

```diff
@@ -46,71 +47,113 @@ export class MarkdownPlugin implements IPlugin<IMarkdownPluginData> {
      * Returns a plain object mapping page references to their data.
      */
     public compile(markdownFiles: IFile[], compiler: ICompiler) {
-        const pageStore = this.buildPageMap(markdownFiles, compiler);
-        // nav must be generated before pages because it rewrites references
-        const nav = this.buildNavTree(pageStore);
-        this.buildPageObject(pageStore, nav);
+        const pageStore = this.buildPageStore(markdownFiles, compiler);
+        // now that we have all known pages, we can resolve @include tags.
+        this.resolveIncludeTags(pageStore);
+        // generate navigation tree after all pages loaded and processed.
+        const nav = pageStore.toTree(this.options.navPage).children as IPageNode[];
+        // use nav tree to fill in `route` for all pages and headings.
+        this.resolveRoutes(pageStore, nav);
+        // generate object at the end, after `route` has been computed throughout.
         const pages = pageStore.toObject();
         return { nav, pages };
     }
 
-    private buildNavTree(pages: PageMap) {
-        return pages.toTree(this.options.navPage).children as IPageNode[];
+    private blockToPage(filePath: string, block: IBlock): IPageData {
+        const reference = getReference(filePath, block);
+        return {
+            reference,
+            route: reference,
+            title: getTitle(block),
+            ...block,
+        };
     }
 
-    private buildPageMap(markdownFiles: IFile[], { renderBlock }: ICompiler) {
-        const pageStore: PageMap = new PageMap();
-        markdownFiles
-            .map((file) => pageStore.add({
-                absolutePath: file.path,
-                ...renderBlock(file.read()),
-            }))
-            .map((page) => {
-                // using `reduce` so we can add one or many entries for each node
-                page.contents = page.contents.reduce((array, content) => {
-                    if (typeof content === "string" || content.tag !== "include") {
-                        return array.concat(content);
-                    }
-                    // inline @include page
-                    const pageToInclude = pageStore.get(content.value);
-                    if (pageToInclude === undefined) {
-                        throw new Error(`Unknown @include reference '${content.value}' in '${page.reference}'`);
-                    }
-                    return array.concat(pageToInclude.contents);
-                }, [] as StringOrTag[]);
-                return page;
-            });
+    /** Convert each file to IPageData and populate store. */
+    private buildPageStore(markdownFiles: IFile[], { renderBlock }: ICompiler) {
+        const pageStore = new PageMap();
```

Comment:
i like map because to me it means there is a string key for each page (which is the case) as opposed to some rando database table of pages.